README
Dist::Zilla::Util::ExpandINI

This module builds upon the previous work ":Util::BundleInfo" ( Which
can extract configuration from a bundle in a manner similar to how
"dzil" does it ) and integrates it with some *very* minimal "INI"
handling to provide a tool capable of generating bundle-free "dist.ini"
files from bundle-using "dist.ini" files!

At present its very naïve and only keeps semantic ordering, and I've
probably gotten something wrong due to cutting the complexity of
Config::MVP out of the loop.

But at this stage, bundles are the *only* thing modified in transit.

Every thing else is practically a token-level copy-paste.

POD ERRORS

Hey! The above document had some coding errors, which are explained
below:

Around line 8:
    Non-ASCII character seen before =encoding in 'naïve'. Assuming
    CP1252

INSTALLATION

This is a Perl module distribution. It should be installed with whichever
tool you use to manage your installation of Perl, e.g. any of

  cpanm .
  cpan  .
  cpanp -i .

Consult http://www.cpan.org/modules/INSTALL.html for further instruction.
Should you wish to install this module manually, the procedure is

  perl Makefile.PL
  make
  make test
  make install

COPYRIGHT AND LICENSE

This software is copyright (c) 2015 by Kent Fredric
<kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Release history for Dist-Zilla-Util-ExpandINI

0.003004 2017-03-04T05:52:50Z e015063

0.003003 2017-03-04T05:26:19Z d66d358
 [Bugfix]
 - Avoid test failures on 5.26 w/ -Ddefault_inc_excludes_dot
 - Closes https://github.com/kentnl/Dist-Zilla-Util-ExpandINI/issues/3

 [Dependencies::Stats]
 - Dependencies changed since 0.003002, see misc/*.deps* for details
 - configure: ↑1 -2 (recommends: -1)
 - develop: +1 ↑1 -1 (suggests: ↑1)
 - test: ↓1

 [Misc]
 - Use newer EUMM w/ ranged versions instead of DynamicPrereqs magic

0.003002 2015-10-19T10:20:25Z 5d62f99
 [Dependencies::Stats]
 - Dependencies changed since 0.003001, see misc/*.deps* for details
 - configure: +2
 - develop: +5 ↑2 -1 (suggests: ↑2)

 [Packaging]
 - Use a simpler DynamicPrereqs invocation, thanks @ether
 - closes #2

0.003001 2015-04-11T05:39:59Z 818763d
 [Dependencies::Stats]
 - Dependencies changed since 0.003000, see misc/*.deps* for details
 - develop: +4 ↑3 -1 (recommends: ↓1, suggests: ↑2)
 - runtime: +1
 - test: (recommends: ↓1)

 [Minor]
 - Packaging and Dependency changes to try avoiding defects w/ Mixin::Linewise 0.107

0.003000 2014-11-08T12:17:58Z 6a88090
 [API]
 - Can now pass `comments => ` with any of the following values to control comment handling.
 - `all`, all comments are copied.
 - `authordeps`, only comments that look like Dist::Zilla AuthorDeps are copied.
 - `none`, no comments are copied.
 - Default is `all`.
 - Thanks to Alan Young ( AYOUNG ) for the contributions ( Github #1 ).

 [Dependencies::Stats]
 - Dependencies changed since 0.002000, see misc/*.deps* for details
 - develop: +1

0.002000 2014-11-08T01:53:39Z da7e0e7
 [API]
 - Now properly instantiable with ->new()
 - Can pass "include_does" and "exclude_does" which take arrays of roles to include/exclude from the results.

 [Dependencies::Stats]
 - Dependencies changed since 0.001003, see misc/*.deps* for details
 - configure: +1 (recommends: ↑1)
 - develop: +7 ↑2 -1 (recommends: +1, suggests: ↑2)
 - runtime: +2 ↓1 -1
 - test: +2 (recommends: ↑2)

 [Features]
 - Comments are preserved in some way, though they may move around a bit.
 - This is mostly to support transferral of authordeps. ( kentnl/Dist-Zilla-App-Command-bakeini#1 )

 [Metadata]
 - author = kentnl

0.001003 2014-08-15T22:30:33Z
 [00 Trivial]
 - CPANDAY!
 - no code changes.
 - tests augmented.
 - whitespace adjusted.

 [Dependencies::Stats]
 - Dependencies changed since 0.001002, see misc/*.deps* for details
 - configure: ↓1
 - develop: +1 (suggests: ↑1)
 - test: -4 (recommends: ↑1 -2)

 [Misc]
 - Improve source side POD to reduce build side whitespace.

 [Tests]
 - update Test::ReportPrereqs

0.001002 2014-07-23T17:44:44Z
 [00 Trivial]
 - Misc packaging cleanups

 [Dependencies::Stats]
 - Dependencies changed since 0.001001, see misc/*.deps* for details
 - develop: +1 (suggests: ↑2)
 - test: +2 ↑1 (recommends: +1)

0.001001 2014-06-03T05:43:02Z
 [Bugfix]
 - use a version of Config::INI with is_valid_property_name

 [Dependencies::Stats]
 - Dependencies changed since 0.001000, see misc/*.deps* for details
 - develop: -1 (suggests: +2 -1)
 - runtime: ↑1

0.001000 2014-06-02T21:24:12Z
 - First version.


maint/perlcritic.rc.gen.pl
#!/usr/bin/env perl
## no critic (Modules::RequireVersionVar)

# ABSTRACT: Write an INI file from a bundle

use 5.008;    # utf8
use strict;
use warnings;
use utf8;

our $VERSION = 0.001;

use Carp qw( croak carp );
use Perl::Critic::ProfileCompiler::Util qw( create_bundle );
use Path::Tiny qw(path);

## no critic (ErrorHandling::RequireUseOfExceptions)
my $bundle = create_bundle('Example::Author::KENTNL');
$bundle->configure;

my @stopwords = (
  qw(
    INI
    naïve
    )
);
for my $wordlist (@stopwords) {
  $bundle->add_or_append_policy_field( 'Documentation::PodSpelling' => ( 'stop_words' => $wordlist ) );
}

#$bundle->add_or_append_policy_field(
#  'Subroutines::ProhibitCallsToUndeclaredSubs' => ( 'exempt_subs' => 'String::Formatter::str_rf' ), );
$bundle->remove_policy('ErrorHandling::RequireUseOfExceptions');
$bundle->remove_policy('CodeLayout::RequireUseUTF8');

#$bundle->remove_policy('ErrorHandling::RequireCarping');
#$bundle->remove_policy('NamingConventions::Capitalization');

my $inf = $bundle->actionlist->get_inflated;

my $config = $inf->apply_config;

{
  my $rcfile = path('./perlcritic.rc')->openw_utf8;
  $rcfile->print( $config->as_ini, "\n" );
  close $rcfile or croak 'Something fubared closing perlcritic.rc';
}
my $deps = $inf->own_deps;
{
  my $target = path('./misc');
  $target->mkpath if not $target->is_dir;

  my $depsfile = $target->child('perlcritic.deps')->openw_utf8;
  for my $key ( sort keys %{$deps} ) {
    $depsfile->printf( "%s~%s\n", $key, $deps->{$key} );
    *STDERR->printf( "%s => %s\n", $key, $deps->{$key} );
  }
  close $depsfile or carp 'Something fubared closing perlcritic.deps';
}



Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.008.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker 7.1101;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Read an INI file and expand bundles as you go.",
  "AUTHOR" => "Kent Fredric <kentnl\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "7.1101"
  },
  "DISTNAME" => "Dist-Zilla-Util-ExpandINI",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "Dist::Zilla::Util::ExpandINI",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Config::INI::Reader" => 0,
    "Config::INI::Writer" => "0.024",
    "Dist::Zilla::Util::BundleInfo" => "1.001000",
    "Mixin::Linewise" => "!= 0.107",
    "Module::Runtime" => 0,
    "Moo" => "1.000008",
    "Scalar::Util" => 0,
    "parent" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Dist::Zilla::PluginBundle::Classic" => 0,
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test::Differences" => 0,
    "Test::More" => 0
  },
  "VERSION" => "0.003004",
  "test" => {
    "TESTS" => "t/*.t t/00-compile/*.t t/comments/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "Config::INI::Reader" => 0,
  "Config::INI::Writer" => "0.024",
  "Dist::Zilla::PluginBundle::Classic" => 0,
  "Dist::Zilla::Util::BundleInfo" => "1.001000",
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Mixin::Linewise" => "!= 0.107",
  "Module::Runtime" => 0,
  "Moo" => "1.000008",
  "Scalar::Util" => 0,
  "Test::Differences" => 0,
  "Test::More" => 0,
  "parent" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
